base/headers.go

```diff
@@ -227,17 +233,115 @@ func (uri WildcardUri) Equals(other Uri) bool {
 }
 
 // Generic list of parameters on a header.
-type Params map[string]MaybeString
+type Params struct {
+	params     map[string]MaybeString
+	paramOrder []string
+}
+
+// Create an empty set of parameters.
+func NewParams() *Params {
+	return &Params{map[string]MaybeString{}, []string{}}
+}
+
+// Returns the entire parameter map.
+func (p *Params) Params() map[string]MaybeString {
+	return p.params
+}
+
+// Returns the requested parameter value.
+func (p *Params) Get(k string) (MaybeString, bool) {
+	v, ok := p.params[k]
+	return v, ok
+}
+
+func (p *Params) Add(k string, v MaybeString) *Params {
+	if p.params == nil {
+		p.params = map[string]MaybeString{}
+		p.paramOrder = []string{}
+	}
+
+	// Add param to order list if new.
+	if _, ok := p.params[k]; !ok {
+		p.paramOrder = append(p.paramOrder, k)
+	}
+
+	// Set param value.
+	p.params[k] = v
+
+	// Return the params so calls can be chained.
+	return p
+}
 
 // Copy a list of params.
-func (p Params) Copy() Params {
-	dup := make(map[string]MaybeString, len(p))
-	for k, v := range p {
-		dup[k] = v
+func (p *Params) Copy() *Params {
+	dup := &Params{}
+	for k, v := range p.params {
```

Comment:
You aren't guaranteeing that you add the elements to the copy in the right order here – map iteration has undefined order.